lib/PuppeteerSharp.Tests/Page/EvaluateTests.cs

```diff
@@ -0,0 +1,166 @@
+                Assert.Equal(42, await frameEvaluation);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldRejectPromiseWithExeption()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var exception = await Assert.ThrowsAsync<EvaluationFailedException>(() =>
+                {
+                    return page.EvaluateAsync<object>("() => not.existing.object.property");
+                });
+
+                Assert.NotNull(exception);
+                Assert.True(exception.Message.Contains("not is not defined"));
+            }
+        }
+
+        [Fact]
+        public async Task SouldReturnComplexObjects()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic obj = new
+                {
+                    foo = "bar!"
+                };
+                dynamic result = await page.EvaluateAsync("a => a", obj);
+                Assert.Equal("bar!", result.foo.ToString());
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNaN()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => NaN");
+                Assert.Equal(double.NaN, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNegative0()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => -0");
+                Assert.Equal(-0, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnInfinity()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => Infinity");
+                Assert.Equal(double.PositiveInfinity, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNegativeInfinty()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => -Infinity");
+                Assert.Equal(double.NegativeInfinity, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldAcceptNullAsOneOfMultipleParameters()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<bool>("(a, b) => Object.is(a, null) && Object.is(b, 'foo')", null, "foo");
+                Assert.True(result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldProperlySerializeNullFields()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => ({a: undefined})");
```

Comment:
shouldn't this serialize to a dictionary and make sure the key `a` exists with a `null` value?

Reply:
I was reviewing this and on Puppeteer's test is returning an empty object https://github.com/GoogleChrome/puppeteer/blob/v1.1.0/test/test.js#L465

I think the name is wrong (even in Puppeteer) this should be called `ShouldProperlyIgnoreUndefinedFields`. I'm creating both tests so we can express the intent here.

Reply:
so i guess we should assert that the object returned is an empty object without properties? this way it looks like it has a null property called `a`.
I would go with something like `EvaluateAsync<Dictionary<string, object>>` and make sure it equals `new Dictionary<string, object>()`

Reply:
Yes, check the updated code ;)